### ISSUE TITLE
Different behavior in OfficeJs (Word) on search/replace between desktop and web application

### ISSUE THREAD

**Issue by mikeappell** [OfficeDev/office-js#2800]
> Message from office-js bot:
We’re closing this issue because it has been inactive for a long time. We’re doing this to keep the issues list manageable and useful for everyone. If this issue is still relevant for you, please create a new issue.
> Thank you for your understanding and continued feedback.

I'm experiencing different behavior when performing a search/replace in MS Word between web and desktop applications.

Note: I originally posted this on StackExchange and it was suggested I create an issue on this repo (see [here](https://stackoverflow.com/questions/73479453/different-behavior-in-officejs-on-search-replace-between-desktop-and-web-applica)).

## Your Environment

* Platform: I'm on MacOS, but others using Windows have experienced the same behavior.
* Host: Word
* Office version number: We're importing `https://appsforoffice.microsoft.com/lib/1.1/hosted/office.js`. Web office version is whatever's current, desktop is 16.64.
* Operating System: Macos/Windows
* Browser: Reproduced on Chrome and Firefox
 
## Expected behavior
The replace function would work correctly in both desktop and web applications.

## Current behavior
When performing a find and replace, _and_ there are multiple instances matching the search pattern found on a single line, _and_ the replaced text is of different length than the original text, then for all replacements after the first one, the text which is replaced is incorrect. It seems like OfficeJS on the web doesn't realize that the positions have changed.

In our actual code, we also encase the replaced text in Content Controls, so I'll include that in the example code below in case it's relevant.

## Steps to reproduce
Create a Word document containing the following text:
```<< One >> << Two >>```
Execute the following code:
```
Word.run((context) => {
  const results = context.document.body.search('[<][<] * [>][>]', { matchWildcards: true });
  context.load(results);

  return context.sync().then(() => {
    results.items.forEach((item) => {
      const text = item.text.replace(/^(<< )/, '').replace(/( >>)$/, '');

      item.insertText(`[${text}]`, Word.InsertLocation.replace);
      const contentControl = item.insertContentControl();
      contentControl.color = 'yellow';
      contentControl.appearance = 'BoundingBox';
      contentControl.removeWhenEdited = true;
    });
  })
});
```
This will work correctly in the Desktop application, resulting in
```[One][Two]```
where each are content controls. However, in Word on the web, it results in 
```[Today] << N[Now] ```
where the bracketed text is correctly encased in Content Controls but the text which was replaced after the first replacement is incorrect.

## Context
Our addin helps our users turn existing Word documents into templates which can then be filled by an outside service of ours. This functionality is very helpful in that regard, and this bug interferes with it, though the fact that it works as expected in the desktop application mitigates the damage.

**Comment by JinghuiMS** [comment]
Thanks for reaching us!
A possible workaround might be rewriting the logic to update only the first result, and search again:

```
async function run() {
  // Gets the current selection and changes the font color to red.
  await Word.run(async (context) => {

    let results = context.document.body.search("[<][<] * [>][>]", { matchWildcards: true });
    context.load(results);

    await context.sync();

    do {
      let item = results.items[0];
      const text = item.text.replace(/^(<< )/, "").replace(/( >>)$/, "");
      item.insertText(`[${text}]`, Word.InsertLocation.replace);
      await context.sync();
      console.log(item);
      item.load();
      await context.sync();
      console.log(item);
      const contentControl = item.insertContentControl();
      contentControl.color = "yellow";
      contentControl.appearance = "BoundingBox";
      contentControl.removeWhenEdited = true;
      await context.sync();
      results = context.document.body.search("[<][<] * [>][>]", { matchWildcards: true });
      context.load(results);

      await context.sync();
    } while (results.items.length > 0)
  });

```
BTW, we have created an internal workitem<ADO#6542408> on it. We will keep track of this issue and let you know if there's any updates.

**Comment by mikeappell** [comment]
Hey @JinghuiMS! Wanted to say thanks for taking on/investigating/assigning this issue.

Re: your workaround suggestion, I'd had a similar thought but was a little concerned about the potential performance impact. In any given document, there's really no way to know how many instances needing to be replaced there will be, either in total or in a specific line.

Could you hazard a guess as to what the impact would be, compared to the current implementation, if there's several lines in the document each with numerous items matching this pattern?

**Comment by JinghuiMS** [comment]
@mikeappell 
We agree with your concern, we are seriously considering a perfect solution, this workaround is only to temporarily unblock you, we agree that this is not a perfect solution, so we created an internal workitem to track this issue.

**Comment by meisammofidi** [comment]
@JinghuiMS 
Would you add that "attention" label to this bug??!! Or any label leads to speed up the process to fix this bug. It seems that it may even take up to a decade to resolve this error duo to this bug is a known issue since 2016. Below you can see the link of proof
https://stackoverflow.com/questions/39787975/office-365-api-search-replace-on-word-online-fault#_=_

I am wondering why this issue is still in force while [Juan Balmori](https://stackoverflow.com/users/5904541/juan-balmori) reported as known issue and its in process to be fixed. It is almost 6 years in process !!!!

I am facing same issue too, `context.document.body.search()` returns values in Word desktop whereas in Word online returns empty object. It would be nice if that issue priorities as high fix in backlog to be fixed and live soon.
